### PR TITLE
Create Initial API Documentation

### PR DESCRIPTION
**Pull request by dogemos**
**Complete**
* Query for resolver / registrar

**TBD**
* Execute messages
* Query for Name NFT contracts

### CODE REVIEW

File: docs/API.md [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+Introduction
+------------
+
+Welcome to the ICNS API documentation! The ICNS API allows you to interact with the Interchain Nameservice, a decentralized protocol that enables users to register and manage human-readable names on the Interchain.
+
+Contract Address
+----------------
+
+*   Name NFT
+    *   Mainnet: TBD
```
Comment (sunnya97) [inline docs/API.md:12]:
```suggestion
    *   Mainnet:  `osmo1xk0s8xgktn9x5vwcgtjdxqzadg88fgn33p8u9cnpdxwemvxscvast52cdd`
```

File: docs/API.md [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+*   Resolver
+    *   Mainnet: TBD
+*   Registrar
+    *   Mainnet: TBD
```
Comment (sunnya97) [inline docs/API.md:16]:
```suggestion
    *   Mainnet:  `osmo1llh07xn7pcst3jqm0xpsucf90lzugfskkkhk8a3u2yznqmse4l5smydwpw`
```

File: docs/API.md [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+Welcome to the ICNS API documentation! The ICNS API allows you to interact with the Interchain Nameservice, a decentralized protocol that enables users to register and manage human-readable names on the Interchain.
+
+Contract Address
+----------------
+
+*   Name NFT
+    *   Mainnet: TBD
+*   Resolver
+    *   Mainnet: TBD
```
Comment (sunnya97) [inline docs/API.md:14]:
```suggestion
    *   Mainnet:  `osmo1mypljhatv0prfr9cjzzvamxdf2ctg34xkt50sudxads9zhqnyneqjuvy26`
```

File: docs/API.md [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+Examples
+--------
+
+**Get a bech32 address from an ICNS name**
+
+```plain
+{endpoint}/cosmwasm/v1/contract/{resolver_contract_address}/smart/{base64Encoding({ "address-by-icns": { "icns": {twitter_username}.{bech32_prefix} }})}
+```
+
+*   Example: [https://lcd.testnet.osmosis.zone/cosmwasm/wasm/v1/contract/osmo1qjs29tsavxuk39dugs7lny7cc28a57hqa9u9q0ysr2qat96aghrqzgp60x/smart/eyJpY25zX25hbWVzIjogeyJhZGRyZXNzIjogIm9zbW8xdnY2aHJ1cXV6cHR5NHhwa3M5em5rdzhneXM1eDRuc25nNGtlcnkifX0K](https://lcd.testnet.osmosis.zone/cosmwasm/wasm/v1/contract/osmo1qjs29tsavxuk39dugs7lny7cc28a57hqa9u9q0ysr2qat96aghrqzgp60x/smart/eyJpY25zX25hbWVzIjogeyJhZGRyZXNzIjogIm9zbW8xdnY2aHJ1cXV6cHR5NHhwa3M5em5rdzhneXM1eDRuc25nNGtlcnkifX0K)
```
Comment (sunnya97) [inline docs/API.md:170]:
```suggestion
*   Example: [https://lcd.osmosis.zone/cosmwasm/wasm/v1/contract/osmo1qjs29tsavxuk39dugs7lny7cc28a57hqa9u9q0ysr2qat96aghrqzgp60x/smart/eyJpY25zX25hbWVzIjogeyJhZGRyZXNzIjogIm9zbW8xdnY2aHJ1cXV6cHR5NHhwa3M5em5rdzhneXM1eDRuc25nNGtlcnkifX0K](https://lcd.osmosis.zone//cosmwasm/wasm/v1/contract/osmo1qjs29tsavxuk39dugs7lny7cc28a57hqa9u9q0ysr2qat96aghrqzgp60x/smart/eyJpY25zX25hbWVzIjogeyJhZGRyZXNzIjogIm9zbW8xdnY2aHJ1cXV6cHR5NHhwa3M5em5rdzhneXM1eDRuc25nNGtlcnkifX0K)
```

File: docs/API.md [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+
+  
+
+**Get a ICNS name of a bech32 address**
+
+```plain
+{endpoint}/cosmwasm/v1/contract/{resolver_contract_address}/smart/{base64Encoding({ "address-by-icns": { "icns": {twitter_username}.{bech32_prefix} }})}
+```
+
+*   Example: [https://lcd.testnet.osmosis.zone/cosmwasm/wasm/v1/contract/osmo1qjs29tsavxuk39dugs7lny7cc28a57hqa9u9q0ysr2qat96aghrqzgp60x/smart/eyJhZGRyZXNzIjogeyJuYW1lIjogIlRvbnlZdW45IiwgImJlY2gzMl9wcmVmaXgiOiAib3NtbyJ9fQo=](https://lcd.testnet.osmosis.zone/cosmwasm/wasm/v1/contract/osmo1qjs29tsavxuk39dugs7lny7cc28a57hqa9u9q0ysr2qat96aghrqzgp60x/smart/eyJhZGRyZXNzIjogeyJuYW1lIjogIlRvbnlZdW45IiwgImJlY2gzMl9wcmVmaXgiOiAib3NtbyJ9fQo=)
```
Comment (sunnya97) [inline docs/API.md:187]:
```suggestion
*   Example: [https://lcd.osmosis.zone/cosmwasm/wasm/v1/contract/osmo1qjs29tsavxuk39dugs7lny7cc28a57hqa9u9q0ysr2qat96aghrqzgp60x/smart/eyJhZGRyZXNzIjogeyJuYW1lIjogIlRvbnlZdW45IiwgImJlY2gzMl9wcmVmaXgiOiAib3NtbyJ9fQo=](https://lcd.osmosis.zone/cosmwasm/wasm/v1/contract/osmo1qjs29tsavxuk39dugs7lny7cc28a57hqa9u9q0ysr2qat96aghrqzgp60x/smart/eyJhZGRyZXNzIjogeyJuYW1lIjogIlRvbnlZdW45IiwgImJlY2gzMl9wcmVmaXgiOiAib3NtbyJ9fQo=)
```

File: docs/API.md (hunk 1 above)
Comment (Nostradamus411) [reply docs/API.md:12]: Sunny got the Name NFT & Resolver contracts flip-flopped

image

```suggestion
    *   Mainnet: `osmo1mypljhatv0prfr9cjzzvamxdf2ctg34xkt50sudxads9zhqnyneqjuvy26`
```

File: docs/API.md (hunk 3 above)
Comment (Nostradamus411) [reply docs/API.md:14]: Sunny got the Name NFT & Resolver contracts flip-flopped

image


```suggestion
    *   Mainnet: `osmo1xk0s8xgktn9x5vwcgtjdxqzadg88fgn33p8u9cnpdxwemvxscvast52cdd`
```

File: docs/API.md (hunk 4 above)
Comment (Nostradamus411) [reply docs/API.md:170]: In addition to the endpoint update to mainnet, the Contract address needs updated to mainnet version and the query was an `icns_names` one so updated with a `address_by_icns` one.
```suggestion
*   Example: [https://lcd.osmosis.zone/cosmwasm/wasm/v1/contract/osmo1xk0s8xgktn9x5vwcgtjdxqzadg88fgn33p8u9cnpdxwemvxscvast52cdd/smart/eyJhZGRyZXNzX2J5X2ljbnMiOiB7ImljbnMiOiAiTm9zTm9kZV8ub3NtbyJ9fQo=](https://lcd.osmosis.zone/cosmwasm/wasm/v1/contract/osmo1xk0s8xgktn9x5vwcgtjdxqzadg88fgn33p8u9cnpdxwemvxscvast52cdd/smart/eyJhZGRyZXNzX2J5X2ljbnMiOiB7ImljbnMiOiAiTm9zTm9kZV8ub3NtbyJ9fQo=)
```

File: docs/API.md [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+*   Query - all\_tokens
+*   Query - minter
+
+Examples
+--------
+
+**Get a bech32 address from an ICNS name**
+
+```plain
+{endpoint}/cosmwasm/v1/contract/{resolver_contract_address}/smart/{base64Encoding({ "address-by-icns": { "icns": {twitter_username}.{bech32_prefix} }})}
```
Comment (Nostradamus411) [inline docs/API.md:167]: Dashes to underscores.
```suggestion
{endpoint}/cosmwasm/v1/contract/{resolver_contract_address}/smart/{base64Encoding({ "address_by_icns": { "icns": {twitter_username}.{bech32_prefix} }})}
```

File: docs/API.md [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+
+```plain
+{
+  "data": {"names":["TonyYun9.osmo"],"primary_name":"TonyYun9.osmo"}
+}
+```
+
+  
+
+**Get a ICNS name of a bech32 address**
```
Comment (Nostradamus411) [inline docs/API.md:181]:
```suggestion
**Get all ICNS names associated with a bech32 address**
```

File: docs/API.md [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+
+```plain
+{endpoint}/cosmwasm/v1/contract/{resolver_contract_address}/smart/{base64Encoding({ "address-by-icns": { "icns": {twitter_username}.{bech32_prefix} }})}
```
Comment (Nostradamus411) [inline docs/API.md:184]:
```suggestion
{endpoint}/cosmwasm/v1/contract/{resolver_contract_address)/smart/{base64Encoding({ "names": { "{bech32Address}" }})}
```

File: docs/API.md (hunk 5 above)
Comment (Nostradamus411) [reply docs/API.md:187]: In addition to the endpoint update to mainnet, the Contract address needs updated to mainnet version and the query was an `icns_names` one so updated with a `names` one.
```suggestion
*   Example: [https://lcd.osmosis.zone/cosmwasm/wasm/v1/contract/osmo1xk0s8xgktn9x5vwcgtjdxqzadg88fgn33p8u9cnpdxwemvxscvast52cdd/smart/eyJuYW1lcyI6IHsiYWRkcmVzcyI6ICJvc21vMTBqbThmdmR5cWxqNzh3MGo1bmF3Yzc2d3NuNHBxbWR4ajRxNXpsIn19Cg==](https://lcd.osmosis.zone/cosmwasm/wasm/v1/contract/osmo1xk0s8xgktn9x5vwcgtjdxqzadg88fgn33p8u9cnpdxwemvxscvast52cdd/smart/eyJuYW1lcyI6IHsiYWRkcmVzcyI6ICJvc21vMTBqbThmdmR5cWxqNzh3MGo1bmF3Yzc2d3NuNHBxbWR4ajRxNXpsIn19Cg==)
```

File: docs/API.md [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+```
+
+*   Example: [https://lcd.testnet.osmosis.zone/cosmwasm/wasm/v1/contract/osmo1qjs29tsavxuk39dugs7lny7cc28a57hqa9u9q0ysr2qat96aghrqzgp60x/smart/eyJhZGRyZXNzIjogeyJuYW1lIjogIlRvbnlZdW45IiwgImJlY2gzMl9wcmVmaXgiOiAib3NtbyJ9fQo=](https://lcd.testnet.osmosis.zone/cosmwasm/wasm/v1/contract/osmo1qjs29tsavxuk39dugs7lny7cc28a57hqa9u9q0ysr2qat96aghrqzgp60x/smart/eyJhZGRyZXNzIjogeyJuYW1lIjogIlRvbnlZdW45IiwgImJlY2gzMl9wcmVmaXgiOiAib3NtbyJ9fQo=)
+*   Response
+
+```plain
+{
+  "data": {"address":"osmo1vv6hruquzpty4xpks9znkw8gys5x4nsng4kery"}
```
Comment (Nostradamus411) [inline docs/API.md:192]: Updating the response to reflect the new example in https://github.com/icns-xyz/icns/pull/19#discussion_r1089647680
```suggestion
  "data": {"names": ["NosNode_", "NosNode_Intern"], "primary_name": "NosNode_"}
```

File: docs/API.md [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+API Endpoints
+-------------
+
+### Resolver APIs
+
+*   Query - config
+    *   Gets the config of the resolver contract, including the Name-NFT contract address.
+    
+    ```plain
+    {endpoint}/cosmwasm/v1/contract/{resolver_contract_address)/smart/{base64Encoding({ "config": { }})}
```
Comment (Nostradamus411) [inline docs/API.md:27]:
```suggestion
    {endpoint}/cosmwasm/v1/contract/{resolver_contract_address}/smart/{base64Encoding({ "config": { }})}
```

File: docs/API.md [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+    ```
+    
+*   Query - addresses\_response
+    *   Returns a list of (bech32\_prefix, bech32\_address) associated with the given ICNS name
+    
+    ```plain
+    {endpoint}/cosmwasm/v1/contract/{resolver_contract_address)/smart/{base64Encoding({ "addresses": { "name": "{icns_name}" })}
```
Comment (Nostradamus411) [inline docs/API.md:34]:
```suggestion
    {endpoint}/cosmwasm/v1/contract/{resolver_contract_address}/smart/{base64Encoding({ "addresses": { "name": "{icns_name}" })}
```

File: docs/API.md [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+    ```
+    
+*   Query - address
+    *   Gets the address associated with the given an ICNS name
+    
+    ```plain
+    {endpoint}/cosmwasm/v1/contract/{resolver_contract_address}/smart/{base64Encoding({ "icns_names": { "address": "{bech32Address}" }})}
```
Comment (Nostradamus411) [inline docs/API.md:41]:
```suggestion
    {endpoint}/cosmwasm/v1/contract/{resolver_contract_address}/smart/{base64Encoding({ "address": { "name": "{icns_name}", "bech32_prefix": "{bech32_prefix}" }})} 
```

File: docs/API.md [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+    {endpoint}/cosmwasm/v1/contract/{resolver_contract_address}/smart/{base64Encoding({ "icns_names": { "address": "{bech32Address}" }})} 
+    ```
+    
+    *   Note: you can register multiple addresses to a name. check for `name` and `primary_name`
+    *   Method to modify the `primary_name` is implemented on the contract, but not on the frontend (future feature)
+*   Query - names
+    *   Returns a list of names associated with the given address
+    
+    ```plain
+    {endpoint}/cosmwasm/v1/contract/{resolver_contract_address)/smart/{base64Encoding({ "names": { "{bech32Address}" }})}
```
Comment (Nostradamus411) [inline docs/API.md:50]:
```suggestion
    {endpoint}/cosmwasm/v1/contract/{resolver_contract_address}/smart/{base64Encoding({ "names": { "{bech32Address}" }})}
```

File: docs/API.md [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+    ```
+    
+    *   NOTE: this query only returns name itself, not full icns name
+        *   does not include ICNS bech32 suffix
+*   Query - icns\_names
+    *   Returns a list of all ICNS names and bech32 prefixes associated with the given address
+    
+    ```plain
+    {endpoint}/cosmwasm/v1/contract/{resolver_contract_address)/smart/{base64Encoding({ "icns_names": { "{bech32Address}" }})}
```
Comment (Nostradamus411) [inline docs/API.md:59]:
```suggestion
    {endpoint}/cosmwasm/v1/contract/{resolver_contract_address}/smart/{base64Encoding({ "icns_names": { "{bech32Address}" }})}
```

File: docs/API.md [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+    ```
+    
+*   Query - primary\_name
+    *   Returns the primary ICNS associated with the given address
+    
+    ```plain
+    {endpoint}/cosmwasm/v1/contract/{resolver_contract_address)/smart/{base64Encoding({ "primary_name": { "{bech32Address}" }})}
```
Comment (Nostradamus411) [inline docs/API.md:66]:
```suggestion
    {endpoint}/cosmwasm/v1/contract/{resolver_contract_address}/smart/{base64Encoding({ "primary_name": { "address": "{bech32Address}" }})}
```

File: docs/API.md [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+    ```
+    
+*   Query - admin
+    *   Returns a list of admins queried from the Name NFT contract.
+    
+    ```plain
+    {endpoint}/cosmwasm/v1/contract/{resolver_contract_address)/smart/{base64Encoding({ "admin": { }})}
```
Comment (Nostradamus411) [inline docs/API.md:73]:
```suggestion
    {endpoint}/cosmwasm/v1/contract/{resolver_contract_address}/smart/{base64Encoding({ "admin": { }})}
```

File: docs/API.md [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+    ```
+    
+*   Query - address\_by\_icns
+    *   Returns the bech32 addresses for the associated full (including the bech32 prefix) ICNS name.
+    
+    ```plain
+    {endpoint}/cosmwasm/v1/contract/{resolver_contract_address}/smart/{base64Encoding({ "address-by-icns": { "icns": {twitter_username}.{bech32_prefix} }})}
```
Comment (Nostradamus411) [inline docs/API.md:80]:
```suggestion
    {endpoint}/cosmwasm/v1/contract/{resolver_contract_address}/smart/{base64Encoding({ "address_by_icns": { "icns": "{twitter_username}.{bech32_prefix}" }})}
```

**Pull request merged**